MLA stack upgrade issues

### What happened?

During MLA stack upgrade from MLA [v0.2.5](https://github.com/kubermatic/mla/releases/tag/v0.2.5) (from the external repo) to the version embedded with KKP v2.22.0-alpha.0 the following issues have found:

Consul upgrade:
```
INFO[0010]    📦 Deploying Consul…
INFO[0010]       Updating release from 0.36.0 to 2.22.0-alpha.0…
ERRO[0015] ❌ Operation failed: failed to deploy Consul: failed to deploy Helm release: failed to install:
Error: UPGRADE FAILED: release consul failed, and has been rolled back due to atomic being set: cannot patch "consul-consul-server" with kind StatefulSet: StatefulSet.apps "consul-consul-server" is invalid: spec: Forbidden: updates to statefulset spec for fields other than 'replicas', 'template', 'updateStrategy', 'persistentVolumeClaimRetentionPolicy' and 'minReadySeconds' are forbidden.
```

Loki upgrade:
```
INFO[0589]    📦 Deploying Loki…
INFO[0589]       Updating release from 0.39.0 to 2.22.0-alpha.0…
ERRO[1567] ❌ Operation failed: failed to deploy Loki: failed to deploy Helm release: failed to install: 
coalesce.go:220: warning: cannot overwrite table with non table for loki-distributed.loki-distributed.querier.affinity (map[])
coalesce.go:220: warning: cannot overwrite table with non table for loki-distributed.loki-distributed.querier.affinity (map[])
coalesce.go:220: warning: cannot overwrite table with non table for loki-distributed.loki-distributed.querier.affinity (map[])
W0207 09:29:18.132882   69113 warnings.go:70] policy/v1beta1 PodDisruptionBudget is deprecated in v1.21+, unavailable in v1.25+; use policy/v1 PodDisruptionBudget
W0207 09:29:18.193633   69113 warnings.go:70] policy/v1beta1 PodDisruptionBudget is deprecated in v1.21+, unavailable in v1.25+; use policy/v1 PodDisruptionBudget
W0207 09:29:18.242387   69113 warnings.go:70] policy/v1beta1 PodDisruptionBudget is deprecated in v1.21+, unavailable in v1.25+; use policy/v1 PodDisruptionBudget
W0207 09:29:18.273288   69113 warnings.go:70] policy/v1beta1 PodDisruptionBudget is deprecated in v1.21+, unavailable in v1.25+; use policy/v1 PodDisruptionBudget
W0207 09:29:18.306733   69113 warnings.go:70] policy/v1beta1 PodDisruptionBudget is deprecated in v1.21+, unavailable in v1.25+; use policy/v1 PodDisruptionBudget
W0207 09:29:18.465820   69113 warnings.go:70] policy/v1beta1 PodDisruptionBudget is deprecated in v1.21+, unavailable in v1.25+; use policy/v1 PodDisruptionBudget
Error: UPGRADE FAILED: release loki-distributed failed, and has been rolled back due to atomic being set: client rate limiter Wait returned an error: rate: Wait(n=1) would exceed context deadline.
```

Several loki pods (compactor, distributor, ingester, querier, ...) stuck in crashloop, with the following error in the logs:
```
level=error ts=2023-02-07T08:15:11.662677578Z caller=main.go:51 msg="validating config" err="invalid queryrange config: the yaml flag `split_queries_by_interval` must now be set in the `limits_config` section instead of the `query_range` config section"
```

More details can be found in this comment: https://github.com/kubermatic/kubermatic/issues/11753#issuecomment-1420447011

### Expected behavior

The MLA stack upgrade should be successful on first attempt.

--- a/cmd/kubermatic-installer/cmd_deploy.go
+++ b/cmd/kubermatic-installer/cmd_deploy.go
@@ -48,7 +48,8 @@ import (
 )
 
 var (
-	MinHelmVersion = semverlib.MustParse("v3.0.0")
+	MinHelmVersion            = semverlib.MustParse("v3.0.0")
+	UserClusterMinHelmTimeout = 15 * time.Minute
 )
 
 type DeployOptions struct {
@@ -148,6 +149,15 @@ func DeployFunc(logger *logrus.Logger, versions kubermaticversion.Versions, opt 
 			fields["git"] = versions.KubermaticCommit
 		}
 
+		stackName := ""
+		if len(args) > 0 {
+			stackName = args[0]
+		}
+		if stackName == "usercluster-mla" && opt.HelmTimeout <= UserClusterMinHelmTimeout {
+			logger.Infof("🚦️ For usercluster-mla deployment, it is recommended to use Helm timeout value of at least %v. Overriding the current value of %s.", UserClusterMinHelmTimeout, opt.HelmTimeout)
+			opt.HelmTimeout = UserClusterMinHelmTimeout
+		}
+
 		// error out early if there is no useful Helm binary
 		helmClient, err := helm.NewCLI(opt.HelmBinary, opt.Kubeconfig, opt.KubeContext, opt.HelmTimeout, logger)
 		if err != nil {
@@ -166,11 +176,6 @@ func DeployFunc(logger *logrus.Logger, versions kubermaticversion.Versions, opt 
 				opt.HelmBinary,
 				helmVersion,
 			)
-		}
-
-		stackName := ""
-		if len(args) > 0 {
-			stackName = args[0]
 		}
 
 		var kubermaticStack stack.Stack

--- a/pkg/install/stack/usercluster-mla/stack.go
+++ b/pkg/install/stack/usercluster-mla/stack.go
@@ -91,6 +91,8 @@ func (*UserClusterMLA) Name() string {
 }
 
 func (s *UserClusterMLA) Deploy(ctx context.Context, opt stack.DeployOptions) error {
+	opt.Logger.Info("ℹ️ For usercluster-mla deployment, it is recommended to extend helm timeout by using flag --helm-timeout=15m0s...")
+
 	if err := deployMLASecrets(ctx, opt.Logger, opt.KubeClient, opt.HelmClient, opt); err != nil {
 		return fmt.Errorf("failed to deploy MLA Secrets: %w", err)
 	}
@@ -209,6 +211,16 @@ func deployConsul(ctx context.Context, logger *logrus.Entry, kubeClient ctrlrunt
 	if err != nil {
 		return fmt.Errorf("failed to check to Helm release: %w", err)
 	}
+	v22 := semverlib.MustParse("2.22.0")
+
+	if release != nil && release.Version.LessThan(v22) && !chart.Version.LessThan(v22) {
+		sublogger.Warn("Installation process will temporarily remove and then upgrade Statefulset used by Consul.")
+
+		err = upgradeConsulStatefulsets(ctx, sublogger, kubeClient, helmClient, opt, chart, release)
+		if err != nil {
+			return fmt.Errorf("failed to prepare Consul for upgrade: %w", err)
+		}
+	}
 
 	if err := util.DeployHelmChart(ctx, sublogger, helmClient, chart, ConsulNamespace, ConsulReleaseName, opt.HelmValues, true, opt.ForceHelmReleaseUpgrade, opt.DisableDependencyUpdate, release); err != nil {
 		return fmt.Errorf("failed to deploy Helm release: %w", err)
@@ -244,7 +256,7 @@ func deployCortex(ctx context.Context, logger *logrus.Entry, kubeClient ctrlrunt
 
 		err = upgradeCortexStatefulsets(ctx, sublogger, kubeClient, helmClient, opt, chart, release)
 		if err != nil {
-			return fmt.Errorf("failed to prepare nginx-ingress-controller for upgrade: %w", err)
+			return fmt.Errorf("failed to prepare Cortex for upgrade: %w", err)
 		}
 	}
 
@@ -413,7 +425,7 @@ func upgradeCortexStatefulsets(
 ) error {
 	logger.Infof("%s: %s detected, performing upgrade to %s…", release.Name, release.Version.String(), chart.Version.String())
 	// 1: find the old deployment
-	logger.Info("Backing up old ingress deployment…")
+	logger.Info("Backing up old memcached statefulsets…")
 
 	statefulsetsList := &unstructured.UnstructuredList{}
 	statefulsetsList.SetGroupVersionKind(schema.GroupVersionKind{
@@ -479,4 +491,51 @@ func getMemcachedStatefulsets(
 		return nil, fmt.Errorf("Error querying API for the existing Deployment object, aborting upgrade process.")
 	}
 	return statefulsetsList, nil
+}
+
+func upgradeConsulStatefulsets(
+	ctx context.Context,
+	logger *logrus.Entry,
+	kubeClient ctrlruntimeclient.Client,
+	helmClient helm.Client,
+	opt stack.DeployOptions,
+	chart *helm.Chart,
+	release *helm.Release,
+) error {
+	logger.Infof("%s: %s detected, performing upgrade to %s…", release.Name, release.Version.String(), chart.Version.String())
+	// 1: find the old deployment
+	logger.Info("Backing up old consul statefulset…")
+
+	statefulsetsList := &unstructured.UnstructuredList{}
+	statefulsetsList.SetGroupVersionKind(schema.GroupVersionKind{
+		Group:   "apps",
+		Kind:    "StatefulSetList",
+		Version: "v1",
+	})
+
+	consulMatcher := ctrlruntimeclient.MatchingLabels{
+		"app":                          ConsulReleaseName,
+		"app.kubernetes.io/managed-by": "Helm",
+		"release":                      release.Name,
+	}
+	if err := kubeClient.List(ctx, statefulsetsList, ctrlruntimeclient.InNamespace(ConsulNamespace), consulMatcher); err != nil {
+		return fmt.Errorf("Error querying API for the existing StatefulSet object, aborting upgrade process.")
+	}
+
+	// 2: store the deployment for backup
+	backupTS := time.Now().Format("2006-01-02T150405")
+	filename := fmt.Sprintf("backup_%s_%s.yaml", ConsulReleaseName, backupTS)
+	logger.Infof("Attempting to store the statefulsets in file: %s", filename)
+	if err := util.DumpResources(ctx, filename, statefulsetsList.Items); err != nil {
+		return fmt.Errorf("failed to back up the statefulsets, it is not removed: %w", err)
+	}
+
+	// 3: delete the deployment
+	logger.Info("Deleting the statefulsets from the cluster")
+	for _, obj := range statefulsetsList.Items {
+		if err := kubeClient.Delete(ctx, &obj); err != nil {
+			return fmt.Errorf("failed to remove the statefulset: %w\n\nuse backup file to check the changes and restore if needed", err)
+		}
+	}
+	return nil
 }

--- a/pkg/install/stack/usercluster-mla/stack.go
+++ b/pkg/install/stack/usercluster-mla/stack.go
@@ -91,7 +91,7 @@ func (*UserClusterMLA) Name() string {
 }
 
 func (s *UserClusterMLA) Deploy(ctx context.Context, opt stack.DeployOptions) error {
-	opt.Logger.Info("ℹ️ For usercluster-mla deployment, it is recommended to extend helm timeout by using flag --helm-timeout=15m0s...")
+	opt.Logger.Info("ℹ️ For usercluster-mla deployment, it is recommended to extend Helm timeout by using the flag --helm-timeout=15m0s...")
 
 	if err := deployMLASecrets(ctx, opt.Logger, opt.KubeClient, opt.HelmClient, opt); err != nil {
 		return fmt.Errorf("failed to deploy MLA Secrets: %w", err)

--- a/pkg/install/stack/usercluster-mla/stack.go
+++ b/pkg/install/stack/usercluster-mla/stack.go
@@ -91,7 +91,6 @@ func (*UserClusterMLA) Name() string {
 }
 
 func (s *UserClusterMLA) Deploy(ctx context.Context, opt stack.DeployOptions) error {
-
 	if err := deployMLASecrets(ctx, opt.Logger, opt.KubeClient, opt.HelmClient, opt); err != nil {
 		return fmt.Errorf("failed to deploy MLA Secrets: %w", err)
 	}

--- a/pkg/install/stack/usercluster-mla/stack.go
+++ b/pkg/install/stack/usercluster-mla/stack.go
@@ -91,7 +91,6 @@ func (*UserClusterMLA) Name() string {
 }
 
 func (s *UserClusterMLA) Deploy(ctx context.Context, opt stack.DeployOptions) error {
-	opt.Logger.Info("ℹ️ For usercluster-mla deployment, it is recommended to extend Helm timeout by using the flag --helm-timeout=15m0s...")
 
 	if err := deployMLASecrets(ctx, opt.Logger, opt.KubeClient, opt.HelmClient, opt); err != nil {
 		return fmt.Errorf("failed to deploy MLA Secrets: %w", err)

